// @totvs/thf-ui/components/thf-page/thf-page-detail/thf-page-detail.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewContainerRef } from '@angular/core';
import { ThfPageDetailBaseComponent } from './thf-page-detail-base.component';
import { callAction, hasAction } from '../thf-page-util/thf-page-util';
/**
 * \@docsExtends ThfPageDetailBaseComponent
 *
 * \@example
 *
 * <example name="thf-page-detail-basic" title="Totvs Page Detail Basic">
 *  <file name="sample-thf-page-detail-basic/sample-thf-page-detail-basic.component.html"> </file>
 *  <file name="sample-thf-page-detail-basic/sample-thf-page-detail-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-detail-labs" title="Totvs Page Detail Labs">
 *  <file name="sample-thf-page-detail-labs/sample-thf-page-detail-labs.component.html"> </file>
 *  <file name="sample-thf-page-detail-labs/sample-thf-page-detail-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-detail-user" title="Totvs Page Detail - User">
 *  <file name="sample-thf-page-detail-user/sample-thf-page-detail-user.component.html"> </file>
 *  <file name="sample-thf-page-detail-user/sample-thf-page-detail-user.component.ts"> </file>
 * </example>
 */
var ThfPageDetailComponent = /** @class */ (function (_super) {
    __extends(ThfPageDetailComponent, _super);
    function ThfPageDetailComponent(viewRef) {
        var _this = _super.call(this) || this;
        _this.callActionFn = callAction;
        _this.hasActionFn = hasAction;
        _this.parentContext = viewRef['_view']['component'];
        return _this;
    }
    /**
     * @return {?}
     */
    ThfPageDetailComponent.prototype.hasAnyAction = /**
     * @return {?}
     */
    function () {
        return this.hasActionFn('back', this.parentContext) ||
            this.hasActionFn('edit', this.parentContext) ||
            this.hasActionFn('remove', this.parentContext);
    };
    /**
     * @param {?} property
     * @return {?}
     */
    ThfPageDetailComponent.prototype.hasEditFn = /**
     * @param {?} property
     * @return {?}
     */
    function (property) {
        if (property === 'icon') {
            return this.hasActionFn('edit', this.parentContext) ? '' : 'thf-icon-delete';
        }
        else if (property === 'type') {
            return this.hasActionFn('edit', this.parentContext) ? 'default' : 'primary';
        }
        else {
            return '';
        }
    };
    /**
     * @param {?} property
     * @return {?}
     */
    ThfPageDetailComponent.prototype.hasEditOrRemoveFn = /**
     * @param {?} property
     * @return {?}
     */
    function (property) {
        if (property === 'icon') {
            return this.hasActionFn('edit', this.parentContext) || this.hasActionFn('remove', this.parentContext) ? '' : 'thf-icon-arrow-left';
        }
        else if (property === 'type') {
            return this.hasActionFn('edit', this.parentContext) || this.hasActionFn('remove', this.parentContext) ? 'default' : 'primary';
        }
        else {
            return '';
        }
    };
    /**
     * @return {?}
     */
    ThfPageDetailComponent.prototype.hasPageHeader = /**
     * @return {?}
     */
    function () {
        return !!(this.title || this.hasAnyAction() || (this.breadcrumb && this.breadcrumb.items.length));
    };
    ThfPageDetailComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-page-detail', template: "<thf-page> <!-- HEADER --> <thf-page-header *ngIf=\"hasPageHeader()\" [t-breadcrumb]=\"breadcrumb\" [t-title]=\"title\"> <!-- OPERATIONS --> <div *ngIf=\"hasAnyAction()\" class=\"thf-page-header-actions\"> <thf-button *ngIf=\"hasActionFn('back', parentContext)\" [t-icon]=\"hasEditOrRemoveFn('icon')\" [t-label]=\"literals.back\" [t-type]=\"hasEditOrRemoveFn('type')\" (t-click)=\"callActionFn('back', parentContext)\"> </thf-button> <thf-button *ngIf=\"hasActionFn('remove', parentContext)\" [t-icon]=\"hasEditFn('icon')\" [t-label]=\"literals.remove\" [t-type]=\"hasEditFn('type')\" (t-click)=\"callActionFn('remove', parentContext)\"> </thf-button> <thf-button *ngIf=\"hasActionFn('edit', parentContext)\" t-icon=\"thf-icon-edit\" t-type=\"primary\" [t-label]=\"literals.edit\" (t-click)=\"callActionFn('edit', parentContext)\"> </thf-button> </div> </thf-page-header> <!-- CONTENT --> <thf-page-content> <ng-content> </ng-content> </thf-page-content> </thf-page> "
                },] },
    ];
    /** @nocollapse */
    ThfPageDetailComponent.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    return ThfPageDetailComponent;
}(ThfPageDetailBaseComponent));
export { ThfPageDetailComponent };
if (false) {
    /** @type {?} */
    ThfPageDetailComponent.prototype.callActionFn;
    /** @type {?} */
    ThfPageDetailComponent.prototype.hasActionFn;
    /** @type {?} */
    ThfPageDetailComponent.prototype.parentContext;
}
