// @totvs/thf-ui/components/thf-list/thf-list.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThfListComponent } from './thf-list.component';
import { ThfButtonModule } from '../thf-button/thf-button.module';
import { ThfListItemComponent } from './thf-list-item.component';
import { ThfListItemDetailComponent } from './thf-list-item-detail.component';
import { ThfListItemContentComponent } from './thf-list-item-content.component';
/**
 * \@description
 *
 * Módulo do componente thf-list.
 *
 * > Aviso: Componente DEPRECIADO, utilize o componente [`thf-list-view`](/documentation/thf-list-view).
 */
var ThfListModule = /** @class */ (function () {
    function ThfListModule() {
    }
    ThfListModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        ThfButtonModule
                    ],
                    declarations: [
                        ThfListComponent,
                        ThfListItemComponent,
                        ThfListItemDetailComponent,
                        ThfListItemContentComponent
                    ],
                    exports: [
                        ThfListComponent,
                        ThfListItemComponent,
                        ThfListItemDetailComponent,
                        ThfListItemContentComponent
                    ],
                    providers: [],
                    schemas: []
                },] },
    ];
    return ThfListModule;
}());
export { ThfListModule };
