// @totvs/thf-ui/components/thf-field/thf-input/thf-input.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR, NG_VALIDATORS } from '@angular/forms';
import { ThfInputGeneric } from '../thf-input-generic/thf-input-generic';
/**
 * \@docsExtends ThfInputBaseComponent
 *
 * \@example
 *
 * <example name="thf-input-basic" title="Totvs Input Basic">
 *  <file name="sample-thf-input-basic/sample-thf-input-basic.component.html"> </file>
 *  <file name="sample-thf-input-basic/sample-thf-input-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-input-labs" title="Totvs Input Labs">
 *  <file name="sample-thf-input-labs/sample-thf-input-labs.component.html"> </file>
 *  <file name="sample-thf-input-labs/sample-thf-input-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-input-reactive-form" title="Totvs Input - Reactive Form">
 *  <file name="sample-thf-input-reactive-form/sample-thf-input-reactive-form.component.html"> </file>
 *  <file name="sample-thf-input-reactive-form/sample-thf-input-reactive-form.component.ts"> </file>
 * </example>
 */
var ThfInputComponent = /** @class */ (function (_super) {
    __extends(ThfInputComponent, _super);
    function ThfInputComponent(el) {
        return _super.call(this, el) || this;
    }
    /**
     * @param {?} c
     * @return {?}
     */
    ThfInputComponent.prototype.extraValidation = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return null;
    };
    ThfInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-input', template: "<thf-field-container [t-help]=\"help\" [t-label]=\"label\" [t-opcional]=\"!required\"> <div class=\"thf-field-container-content\"> <div *ngIf=\"icon\" class=\"thf-field-icon-container-left\"> <span class=\"thf-icon thf-field-icon {{ icon }}\" [class.thf-field-icon-disabled]=\"disabled\"></span> </div> <input #inp class=\"thf-input\" [attr.name]=\"name\" [class.thf-input-icon-left]=\"icon\" [class.thf-input-icon-right]=\"clean\" [disabled]=\"disabled\" [placeholder]=\"placeholder\" [readonly]=\"readonly\" [required]=\"required\" [type]=\"type\" (blur)=\"eventOnBlur($event)\" (click)=\"eventOnClick($event)\" (focus)=\"eventOnFocus($event)\" (input)=\"eventOnInput($event)\"> <div class=\"thf-field-icon-container-right\"> <thf-clean [t-element-ref]=\"inputEl\" (t-change-event)=\"clear($event)\"></thf-clean> </div> </div> <thf-field-container-bottom [t-error-pattern]=\"getErrorPattern()\"> </thf-field-container-bottom> </thf-field-container> ",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return ThfInputComponent; }),
                            multi: true,
                        },
                        {
                            provide: NG_VALIDATORS,
                            useExisting: forwardRef(function () { return ThfInputComponent; }),
                            multi: true,
                        }
                    ]
                },] },
    ];
    /** @nocollapse */
    ThfInputComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    return ThfInputComponent;
}(ThfInputGeneric));
export { ThfInputComponent };
