// @totvs/thf-ui/components/thf-field/thf-lookup/thf-lookup-modal/thf-lookup-modal.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, ViewChild } from '@angular/core';
import { fromEvent } from 'rxjs';
import { debounceTime, filter } from 'rxjs/operators';
import { ThfLookupModalBaseComponent } from '../thf-lookup-modal/thf-lookup-modal-base.component';
/**
 * \@docsPrivate
 *
 * \@docsExtends ThfLookupModalBaseComponent
 */
var ThfLookupModalComponent = /** @class */ (function (_super) {
    __extends(ThfLookupModalComponent, _super);
    function ThfLookupModalComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.keyUpObservable = null;
        _this.containerHeight = 375;
        _this.tableHeight = 370;
        return _this;
    }
    /**
     * @return {?}
     */
    ThfLookupModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.initializeEventInput();
        this.setTableHeight();
    };
    /**
     * @return {?}
     */
    ThfLookupModalComponent.prototype.initializeEventInput = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.keyUpObservable = fromEvent(this.inputSearchEl.nativeElement, 'keyup')
            .pipe(filter(function (e) { return _this.validateEnterPressed(e); }), debounceTime(400));
        this.keyUpObservable.subscribe(function () {
            _this.search();
        });
    };
    /**
     * @return {?}
     */
    ThfLookupModalComponent.prototype.openModal = /**
     * @return {?}
     */
    function () {
        this.thfModal.open();
    };
    /**
     * @private
     * @return {?}
     */
    ThfLookupModalComponent.prototype.setTableHeight = /**
     * @private
     * @return {?}
     */
    function () {
        if (window.innerHeight < 615) {
            this.tableHeight -= 50;
            this.containerHeight -= 50;
        }
    };
    /**
     * @private
     * @param {?} e
     * @return {?}
     */
    ThfLookupModalComponent.prototype.validateEnterPressed = /**
     * @private
     * @param {?} e
     * @return {?}
     */
    function (e) {
        return e.keyCode === 13;
    };
    ThfLookupModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-lookup-modal', template: "<thf-modal t-click-out=\"false\" t-hide-close=\"false\" t-size=\"lg\" [t-primary-action]=\"primaryAction\" [t-secondary-action]=\"secondaryAction\" [t-title]=\"title\"> <thf-field-container class=\"thf-lookup-header thf-md-6 thf-pull-right\" [t-opcional]=\"false\"> <div class=\"thf-field-container-content\"> <input #inpsearch class=\"thf-input thf-input-icon-right\" name=\"contentSearch\" [(ngModel)]=\"searchValue\" [placeholder]=\"literals.modalPlaceholder\" type=\"text\"> <div class=\"thf-field-icon-container-right\"> <span #iconLookup class=\"thf-icon thf-field-icon thf-icon-search\" (click)=\"search()\"> </span> </div> </div> </thf-field-container> <div class=\"thf-row thf-lookup-container-table\" [style.height.px]=\"containerHeight\"> <thf-table class=\"thf-md-12\" t-checkbox=\"true\" t-hide-detail=\"true\" t-single-select=\"true\" t-sort=\"true\" [t-columns]=\"columns\" [t-height]=\"tableHeight\" [t-items]=\"items\" [t-literals]=\"tableLiterals\" [t-loading]=\"isLoading\" [t-show-more-disabled]=\"!hasNext\" (t-show-more)=\"showMoreEvent()\"> </thf-table> </div> </thf-modal> "
                },] },
    ];
    ThfLookupModalComponent.propDecorators = {
        inputSearchEl: [{ type: ViewChild, args: ['inpsearch',] }]
    };
    return ThfLookupModalComponent;
}(ThfLookupModalBaseComponent));
export { ThfLookupModalComponent };
if (false) {
    /** @type {?} */
    ThfLookupModalComponent.prototype.inputSearchEl;
    /** @type {?} */
    ThfLookupModalComponent.prototype.keyUpObservable;
    /** @type {?} */
    ThfLookupModalComponent.prototype.containerHeight;
    /** @type {?} */
    ThfLookupModalComponent.prototype.tableHeight;
}
