// @totvs/thf-ui/components/thf-menu-panel/thf-menu-panel.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewContainerRef } from '@angular/core';
import { Location } from '@angular/common';
import { NavigationEnd, Router } from '@angular/router';
import { callFunction, getFormattedLink, openExternalLink } from '../../utils/util';
import { ThfMenuPanelBaseComponent } from './thf-menu-panel-base.component';
import { ThfMenuPanelItemsService } from './services/thf-menu-panel-items.service';
/**
 * \@docsExtends ThfMenuPanelBaseComponent
 *
 * \@description
 *
 * Para o menu funcionar corretamente é necessário importar o `RouterModule` e `Routes` do módulo principal de
 *  sua aplicação:
 *
 * ````
 * import { RouterModule, Routes } from '\@angular/router';
 *
 * ...
 *
 * \@NgModule({
 *   imports: [
 *     RouterModule,
 *     Routes,
 *     ...
 *     ThfModule,
 *     ...
 *   ],
 *   declarations: [
 *     AppComponent
 *   ],
 *   providers: [],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule { }
 * ```
 *
 * Além disso é necessário criar um módulo configurando as rotas da aplicação.
 *
 * ```
 * import { NgModule } from '\@angular/core';
 *
 * import { RouterModule, Routes } from '\@angular/router';
 *
 * import { HelloWorldComponent } from './hello-world/hello-world.component';
 *
 * const routes: Routes = [
 *   {path: 'hello-world', component: HelloWorldComponent}
 * ];
 *
 * \@NgModule({
 *   imports: [RouterModule.forRoot(routes, {useHash: true})],
 *   exports: [RouterModule]
 * })
 * export class AppRoutingModule {}
 * ```
 *
 * \@example
 *
 * <example name="thf-menu-panel-basic" title="Totvs Menu Panel Basic">
 *   <file name="sample-thf-menu-panel-basic/sample-thf-menu-panel-basic.component.html"> </file>
 *   <file name="sample-thf-menu-panel-basic/sample-thf-menu-panel-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-menu-panel-labs" title="Totvs Menu Panel Labs">
 *   <file name="sample-thf-menu-panel-labs/sample-thf-menu-panel-labs.component.html"> </file>
 *   <file name="sample-thf-menu-panel-labs/sample-thf-menu-panel-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-menu-panel-customer" title="Totvs Menu Panel - Customers">
 *   <file name="sample-thf-menu-panel-customer/sample-thf-menu-panel-customer.component.html"> </file>
 *   <file name="sample-thf-menu-panel-customer/sample-thf-menu-panel-customer.component.ts"> </file>
 * </example>
 */
var ThfMenuPanelComponent = /** @class */ (function (_super) {
    __extends(ThfMenuPanelComponent, _super);
    function ThfMenuPanelComponent(viewRef, location, menuItemsService, router) {
        var _this = _super.call(this) || this;
        _this.location = location;
        _this.menuItemsService = menuItemsService;
        _this.router = router;
        _this.parentRef = viewRef['_view']['component'];
        return _this;
    }
    /**
     * @return {?}
     */
    ThfMenuPanelComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.itemSubscription.unsubscribe();
        this.routeSubscription.unsubscribe();
    };
    /**
     * @return {?}
     */
    ThfMenuPanelComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.subscribeToMenuItem();
        this.subscribeToRoute();
    };
    /**
     * @private
     * @param {?} urlPath
     * @param {?} menus
     * @return {?}
     */
    ThfMenuPanelComponent.prototype.activateMenuByUrl = /**
     * @private
     * @param {?} urlPath
     * @param {?} menus
     * @return {?}
     */
    function (urlPath, menus) {
        var _this = this;
        if (menus) {
            return menus.some(function (menu) {
                if (getFormattedLink(menu.link) === urlPath) {
                    _this.activateMenuItem(menu);
                    return true;
                }
            });
        }
    };
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    ThfMenuPanelComponent.prototype.activateMenuItem = /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        this.activeMenuItem = menu;
        this.linkActive = getFormattedLink(menu.link);
        this.menuItemsService.sendToChildMenuClicked({ active: this.activeMenuItem, activatedByRoute: true });
    };
    /**
     * @private
     * @param {?} urlPath
     * @return {?}
     */
    ThfMenuPanelComponent.prototype.checkActiveMenuByUrl = /**
     * @private
     * @param {?} urlPath
     * @return {?}
     */
    function (urlPath) {
        if (!this.linkActive || this.linkActive !== urlPath) {
            this.activateMenuByUrl(urlPath, this.menus);
        }
    };
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    ThfMenuPanelComponent.prototype.clickMenuItem = /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        if (menu.action) {
            callFunction(menu.action, this.parentRef, menu);
        }
        if (menu.type === 'externalLink') {
            openExternalLink(menu.link);
        }
        else if (menu.type === 'internalLink') {
            this.activateMenuItem(menu);
        }
    };
    /**
     * @private
     * @return {?}
     */
    ThfMenuPanelComponent.prototype.subscribeToRoute = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.routeSubscription = this.router.events.subscribe(function (rounterEvent) {
            if (rounterEvent instanceof NavigationEnd) {
                _this.checkActiveMenuByUrl(_this.location.path());
            }
        });
    };
    /**
     * @private
     * @return {?}
     */
    ThfMenuPanelComponent.prototype.subscribeToMenuItem = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.itemSubscription =
            this.menuItemsService.receiveFromChildMenuClicked().subscribe(function (menu) { return _this.clickMenuItem(menu); });
    };
    ThfMenuPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-menu-panel', template: "<div class=\"thf-menu-panel\"> <div class=\"thf-menu-panel-logo-container\"> <a href=\"./\"> <div class=\"thf-menu-panel-logo\"></div> </a> </div> <nav class=\"thf-menu-panel-container\"> <div class=\"thf-menu-panel-inner\"> <div *ngFor=\"let menu of menus, let menuIndex = index;\" class=\"thf-menu-panel-item-wrapper\"> <thf-menu-panel-item [class.thf-menu-panel-item-first]=\"menuIndex === 0\" [t-menu-item-internal]=\"menu\"> </thf-menu-panel-item> </div> </div> </nav> </div> "
                },] },
    ];
    /** @nocollapse */
    ThfMenuPanelComponent.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: Location },
        { type: ThfMenuPanelItemsService },
        { type: Router }
    ]; };
    return ThfMenuPanelComponent;
}(ThfMenuPanelBaseComponent));
export { ThfMenuPanelComponent };
if (false) {
    /** @type {?} */
    ThfMenuPanelComponent.prototype.activeMenuItem;
    /** @type {?} */
    ThfMenuPanelComponent.prototype.linkActive;
    /** @type {?} */
    ThfMenuPanelComponent.prototype.parentRef;
    /**
     * @type {?}
     * @private
     */
    ThfMenuPanelComponent.prototype.routeSubscription;
    /**
     * @type {?}
     * @private
     */
    ThfMenuPanelComponent.prototype.itemSubscription;
    /**
     * @type {?}
     * @private
     */
    ThfMenuPanelComponent.prototype.location;
    /**
     * @type {?}
     * @private
     */
    ThfMenuPanelComponent.prototype.menuItemsService;
    /**
     * @type {?}
     * @private
     */
    ThfMenuPanelComponent.prototype.router;
}
